Hadouken v5 downloader implementation.

vktr opened the pull request:
I've implemented a `DownloadClient` to support getting torrents with [Hadouken](https://github.com/hadouken/hadouken).

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenSettingsValidator.cs:
```diff
@@ -0,0 +1,29 @@
+            RuleFor(c => c.Host).ValidHost();
+            RuleFor(c => c.Port).GreaterThan(0);
+
+            RuleFor(c => c.Token)
```
First rule should be on the first line then align on the `.`.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenResponse.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public class HadoukenResponse<TResult>
+    {
+        public int Id { get; set; }
```
Single line properties should not have a blank line after them.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenProxy.cs:
```diff
@@ -0,0 +1,116 @@
+            ProcessRequest<bool>(settings, "session.removeTorrent", downloadId, deleteData);
+        }
+
+        private HadoukenResponse<TResult> ProcessRequest<TResult>(HadoukenSettings settings,
```
All arguments should be on the same line.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenProxy.cs:
```diff
@@ -0,0 +1,116 @@
+            return client.ExecuteAndValidate<HadoukenResponse<TResult>>(request);
+        }
+
+        private IRestClient BuildClient(HadoukenSettings settings)
```
Use https://github.com/Sonarr/Sonarr/blob/develop/src/NzbDrone.Common/Http/HttpClient.cs instead of RestClient.

Taloth replied:
@markus101 Pretty much all download client implementations use RestClient atm. Kinda unfair to demand HttpClient from a PR.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenProxy.cs:
```diff
@@ -0,0 +1,116 @@
+
+            if (settings.AuthenticationType == (int) AuthenticationType.Basic)
+            {
+                var basicData = Encoding.UTF8.GetBytes(string.Format("{0}:{1}", settings.Username, settings.Password));
```
HttpClient will use credentials from the http request so you don't need to build this by hand,.

Taloth commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenProxy.cs:
```diff
@@ -0,0 +1,114 @@
+
+        public string AddTorrentFile(HadoukenSettings settings, byte[] fileContent)
+        {
+            return ProcessRequest<string>(settings, "session.addTorrentFile", Convert.ToBoolean(fileContent)).Result;
```
Convert.**ToBoolean**(fileContent) ?!?!

vktr replied:
How did that happen :smile:

vktr commented:
@markus101 I've updated the PR according to your comments (and accidentally squashed that commit into the previous), and also fixed the error that @Taloth found.

Let me know if you want me to use the `IHttpClient` instead.

kayone commented on src/NzbDrone.Core/Download/Clients/Hadouken/Hadouken.cs:
```diff
@@ -0,0 +1,188 @@
+            Logger logger)
+            : base(torrentFileInfoReader, httpClient, configService, diskProvider, remotePathMappingService, logger)
+        {
+            if (proxy == null)
```
you don't need to check for missing arguments, they are passed in by the container and if the container can't resolve them it'll throw itself.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/IHadoukenProxy.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public interface IHadoukenProxy
+    {
+        HadoukenSystemInfo GetSystemInfo(HadoukenSettings settings);
```
No blank lines needed here.

Taloth commented:
I suggest you hold on with pushing commits for now, till we can get all comments in.
I don't have time today to finish a full review, just wanted to point at a few things that stood out at first glance.

My concern is if that ToBoolean error creeped in, it means that function was never tested.
This is a big problem, coz we expect the ```I...Proxy``` classes to be implemented and tested with the actual application. They are the thin layer that talks to the outside world. It's thin intentionally coz it means we can mock and unit test on the very edge. But it also means we gotta guarantee that the interface is correct.

Also, you have to populate the item.IsReadOnly flag properly, it's used to determine if the torrent finished seeding. Assuming Hadouken supports something like 'stop-on-ratio'.
The flag is subsequently used by Sonarr to determine whether the torrent can be removed automatically.

You won't have to use ```IHttpClient```, we'll take care of the switch from RestClient to HttpClient later on.

vktr commented:
@Taloth I see. Yes, the `ToBoolean` error was embarrasing - sorry for that. I'll see if I can add some unit tests for the `HadoukenProxy`. As for the `IsReadOnly` flag, Hadouken is built on libtorrent (same as Deluge) so I can definitely make it work what with the seed ratio and all.

v5 is currently unreleased so there's no rush on my end, it's better to get you guys' input on as much as possible and have a high-quality PR.

Thanks for the comments!

Taloth commented:
@vktr The important part is that you actually test against Hadouken itself and verify that the Proxy responds in the expected way. Once you've verified that, we'll know that the Mocker.GetMock<IHadoukenProxy>().Setup(..) stuff that exist in HadoukenFixture mimics reality and means the rest of the tests actually have meaning.

Also if v5 is the first version supported, get a version check in the Hadouken.Test(), unless Hadouken v4 didn't have an api, see Deluge/uTorrent implementations for an example.

aaronchi commented:
I would love to see this merged

AGresvig commented:
@vktr Don't give in! Good work!!

markus101 commented:
Hey @vktr Just wanted to touch base and see if there is anything you need from us on this PR. Has v5 been released?

vktr commented:
Hi, sorry for the radio silence! Sometimes real life gets the upper hand :smile: 

Yes, v5 has been out for a while, however the API has changed a bit since this PR was created. I'm about to update it with the correct API calls.

However I haven't followed the Sonarr development closely. Perhaps it would be easier to recreate the PR on the current development branch if the discusses RestClient/HttpClient changed were made.

Anyways, I'll make sure this PR gets a well-deserved update in the coming days.

markus101 commented:
Understandable, real life has a habit of doing that.

We haven't made any changes to RestClient/HttpClient, so you're fine with sticking with what you have, a rebase on develop to get everything updated should be all you need.

Looking forward to the update, thanks!

phrusher commented:
@markus101  I believe this is ready for review.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenProxy.cs:
```diff
@@ -0,0 +1,185 @@
+        }
+
+        private HadoukenResponse<TResult> ProcessRequest<TResult>(HadoukenSettings settings,
+            string method,
```
Method arguments on a single line.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/Hadouken.cs:
```diff
@@ -0,0 +1,203 @@
+                if (torrent.IsFinished && torrent.State == HadoukenTorrentState.Paused)
+                {
+                    item.IsReadOnly = false;
+                } else
```
`else` on new line

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/Hadouken.cs:
```diff
@@ -0,0 +1,203 @@
+        private readonly IHadoukenProxy _proxy;
+
+        public Hadouken(IHadoukenProxy proxy,
+            ITorrentFileInfoReader torrentFileInfoReader,
```
Align on `I` to the first line.

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/HadoukenProxy.cs:
```diff
@@ -0,0 +1,185 @@
+
+namespace NzbDrone.Core.Download.Clients.Hadouken
+{
+    public sealed class HadoukenProxy : IHadoukenProxy
```
No need to seal the classes (for this one or other cases)

markus101 commented on src/NzbDrone.Core/Download/Clients/Hadouken/AuthenticationType.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        None = 0,
+        Basic,
+        Token
```
Reading the release notes 5.1 dropped support for anonymous and token authentication, I think we should clean this up and set a minimum required version to 5.1 (I assume we can check the version in the API and have a test for that when testing the connection).

markus101 commented:
There are a few more things to address, but beyond that we'll need to get it rebased off the latest develop changes and get the commits squashed and drop that merge commit. Should end up with two commits, one for each author on the PR if we can.

phrusher commented:
I have corrected the issues based on the remarks. However since my work is in a fork of a fork I can't seem to get my changes to show up here unless i send a PR to vktr and then he in turn updates this PR. I had a short talk with him and he currently had no time to update this issue. Any suggestions of how to proceed?

markus101 commented:
@phrusher best thing would be to open up a new PR.

vktr commented:
Hi, give me today and I'll see if we can get this sorted :) Sorry for the
super delay!

On Monday, 25 January 2016, Mark McDowall <notifications@github.com> wrote:

> @phrusher <https://github.com/phrusher> best thing would be to open up a
> new PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Sonarr/Sonarr/pull/678#issuecomment-174379943>.
>

vktr commented:
@markus101 @phrusher 

I've cleaned up the PR and it should be good to go :smile: Sorry for the biggest delay ever.

I don't know why the build fails - if it's something I can do to help let me know!

vktr commented:
@markus101 @phrusher 

Did some magic to rebase the latest changes from @phrusher. Still don't know what's wrong with the build, though.

Sano commented:
Is this PR still stuck in limbo?

Flaxee commented:
Any work happening on this one? Would be great to get Hadouken as a supported download client in Sonarr.

vktr commented:
Hi! I haven't heard anything since Jan 25, so AFAIK this PR is ready to be merged.

markus101 commented:
I pulled it down rebased it and made a few changes to add label support and convert it to the new HTTP client we're using and its in develop now. Thanks for working on this and sorry for the massive delays.
